infra/baseServer.go

```diff
@@ -58,50 +47,25 @@ agentMessaging:
 				fmt.Println("stopped messaging early", len(serv.agentStoppedTalkingMap), len(serv.agentMap))
 				break agentMessaging
 			}
-			select {
-			case uuid := <-serv.agentServerChannel:
-				fmt.Println(uuid, "has stopped talking")
-				serv.agentStoppedTalkingMap[uuid] = struct{}{}
-			default:
-				continue
-			}
+
 		}
 	}
 	serv.agentStoppedTalkingMap = make(map[uuid.UUID]struct{})
 }
 
 func (server *BaseServer[T]) HandleEndOfTurn(iter, round int) {
-	server.waitForMessagingToEnd()
 	server.endAgentListeningSession()
 	fmt.Printf("Iteration %d, Round %d finished.\n", iter, round)
 }
 
 func (server *BaseServer[T]) RunAgentLoop() {}
 
-func (server *BaseServer[T]) SendMessage(msg IMessage, receivers []uuid.UUID) {
+func (server *BaseServer[T]) SendMessage(msg IMessage[T], receivers []uuid.UUID) {
+	//defer server.listeningWaitGroup.Done()
 	for _, receiver := range receivers {
-		select {
-		case server.agentAgentChannelMap[receiver] <- msg:
-		default:
-		}
+		msg.InvokeMessageHandler(server.agentMap[receiver])
```

Comment:
```suggestion
		go msg.InvokeMessageHandler(server.agentMap[receiver])
```

this should be threaded, no? this prevent the user from having to manually call it as a goroutine, hiding functionality (which is a good thing)

Reply:
good idea i changed it